src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs

```diff
@@ -34,15 +41,27 @@ public NodeStateTable<ImmutableArray<TInput>> UpdateStateTable(DriverStateTable.
             // - Added when the previous table was empty
             // - Modified otherwise
 
-            var source = sourceTable.Batch();
+            // update the table
+            var newTable = builder.CreateTableBuilder(previousTable, _name);
+
+            // If this execution is tracking steps, then the source table should have also tracked steps or be the empty table.
+            Debug.Assert(!newTable.TrackIncrementalSteps || (sourceTable.HasTrackedSteps || sourceTable.IsEmpty));
+
+            var stopwatch = SharedStopwatch.StartNew();
 
-            // update the table 
-            var newTable = previousTable.ToBuilder();
-            if (!sourceTable.IsCached || !newTable.TryUseCachedEntries())
+            var batchedSourceEntries = sourceTable.Batch();
+            var sourceValues = batchedSourceEntries.SelectAsArray(sourceEntry => sourceEntry.State != EntryState.Removed, sourceEntry => sourceEntry.Item);
+            var sourceInputs = newTable.TrackIncrementalSteps ? batchedSourceEntries.SelectAsArray(sourceEntry => (sourceEntry.Step!, sourceEntry.OutputIndex)) : default;
+
+            if (previousTable.IsEmpty)
+            {
+                newTable.AddEntry(sourceValues, EntryState.Added, stopwatch.Elapsed, sourceInputs, EntryState.Added);
+            }
```

Comment:
Is this `IsEmpty` case needed or will the `else if` branch successfully handle this case?

Reply:
I think we could use just the `else if` branch, but we'd end up doing quite a bit of extra computation instead of just this check. This `if`/`else if` split exists with the current implementation today. I think @chsienki could comment on his design choice here.